cpp/include/cuspatial/experimental/detail/is_point_in_polygon_kernel.cuh

```diff
@@ -0,0 +1,113 @@
+ * API.
+ */
+template <class Cart2d,
+          class OffsetType,
+          class OffsetIterator,
+          class Cart2dIt,
+          class OffsetItDiffType = typename std::iterator_traits<OffsetIterator>::difference_type,
+          class Cart2dItDiffType = typename std::iterator_traits<Cart2dIt>::difference_type>
+__device__ inline bool is_point_in_polygon(Cart2d const& test_point,
+                                           OffsetType poly_begin,
+                                           OffsetType poly_end,
+                                           OffsetIterator ring_offsets_first,
+                                           OffsetItDiffType const& num_rings,
+                                           Cart2dIt poly_points_first,
+                                           Cart2dItDiffType const& num_poly_points)
+{
+  using T = iterator_vec_base_type<Cart2dIt>;
+
+  bool point_is_within = false;
+  bool is_colinear     = false;
+  // for each ring
+  for (auto ring_idx = poly_begin; ring_idx < poly_end; ring_idx++) {
+    int32_t ring_idx_next = ring_idx + 1;
+    int32_t ring_begin    = ring_offsets_first[ring_idx];
+    int32_t ring_end =
+      (ring_idx_next < num_rings) ? ring_offsets_first[ring_idx_next] : num_poly_points;
+
+    Cart2d b     = poly_points_first[ring_end - 1];
+    bool y0_flag = b.y > test_point.y;
+    bool y1_flag;
+    // for each line segment, including the segment between the last and first vertex
+    for (auto point_idx = ring_begin; point_idx < ring_end; point_idx++) {
+      Cart2d const a = poly_points_first[point_idx];
+      T run          = b.x - a.x;
+      T rise         = b.y - a.y;
+
+      // Points on the line segment are the same, so intersection is impossible.
+      // This is possible because we allow closed or unclosed polygons.
+      if (run == 0.0 && rise == 0.0) continue;
+
+      T rise_to_point = test_point.y - a.y;
+
+      // colinearity test
+      const T EPSILON = 0.0000000000000000000000000000001;
+      T run_to_point  = test_point.x - a.x;
+      T colinearity   = (run * rise_to_point - run_to_point * rise);
+      is_colinear     = colinearity * colinearity < EPSILON;
+      printf("Colinearity: %.60lf\n", colinearity * colinearity);
+      printf("Is colinear? : %s\n", is_colinear ? "True" : "False");
+      if (is_colinear) { break; }
+
+      y1_flag = a.y > test_point.y;
+      printf("Can intersect? : %s\n", y1_flag != y0_flag ? "True" : "False");
+      if (y1_flag != y0_flag) {
+        // Transform the following inequality to avoid division
+        //  test_point.x < (run / rise) * rise_to_point + a.x
+        auto lhs                  = (test_point.x - a.x) * rise;
+        auto rhs                  = run * rise_to_point;
+        const T INTERSECT_EPSILON = 0.00000000000001;
+        printf("lhs - rhs: %.60lf\n", lhs - rhs);
+        if (lhs - INTERSECT_EPSILON < rhs != y1_flag) {
```

Comment:
I don't see why you need to add epsilons everywhere when the only concern was about the new colinearity check because it looks for exact equality. 

My main concern is that the colinearity test is basically doing the same math that the intersection check is already doing, just replacing inequality with equality. This is what led me to believe there must be a formulation of this algorithm which excludes boundaries more naturally.

Reply:
If you have excluded all colinear cases, then the ray crossings test here is non-ambiguous. Introducing extra epsilon here is introducing unnecessary bias into the equation.